worktree_status.go

```diff
@@ -47,8 +47,45 @@ func (w *Worktree) Status() (Status, error) {
 	return w.status(hash)
 }
 
+func (w *Worktree) newStatusFromIndex() (Status, error) {
+	idx, err := w.r.Storer.Index()
+	if err != nil {
+		return nil, err
+	}
+
+	idxRoot := mindex.NewRootNode(idx)
+	nodes := []noder.Noder{idxRoot}
+
+	if err != nil {
+		return nil, err
+	}
+
+	status := make(Status)
+
+	for len(nodes) > 0 {
+		var node noder.Noder
+		node, nodes = nodes[0], nodes[1:]
+		if node.IsDir() {
+			children, err := node.Children()
+			if err != nil {
+				return nil, err
+			}
+			nodes = append(nodes, children...)
+			continue
+		}
+		fs := status.File(node.Name())
+		fs.Worktree = Unmodified
+		fs.Staging = Unmodified
+	}
```

Comment:
This may have a performance impact, as we need to go through all nodes while growing `nodes` as part of that process.

What if instead, we amended `File(path string)` to validate the returning [s[path]](https://github.com/go-git/go-git/blob/master/status.go#L20) and force a default value of `Unmodified` if no value was currently set?

Reply:
Sorry, I didn't understand what you proposed. Can you explain it a little more?

Reply:
The issue here is that going through the entire worktree to create an initial state as unmodified can be quite costly - specially on large repositories. I tested this on a mid-sized repository and noticed that with these changes the `.Status()` call took ~30% longer than before.

My initial suggestion was to amend `Status` so that the `File` func would effectively do the same thing as the pre-init you created. However, this won't really work in some cases, as when there are no changes the `Status` returns an empty list. Besides, the existing logic is incorrect as whatever file we pass on to `status.File(path)` it returns `Untracked` even if that file does not exist across both the staging and the worktree.

In order to fix this, while keeping backwards compatibility and not introducing a performance regression, I think we have a few options:
a) Introduce a new `Worktree.StatusWithOptions()` func which returns a new type that is more flexible and calculates the status of a given file on demand. I suspect this will need to have access to the worktree fs, so that it can handle the scenario I mentioned above, without needing to front-load the entire worktree. It may also need access to the exclusion rules.
b) Introduce a new `Worktree.StatusWithOptions()` func which returns a new type that is more flexible, but simply adds an option so users can opt-in to the front-loading this PR currently introduces.
c) Introduce a new `Worktree.FileStatus()` func which calls `.Status` and ensure the given file is correctly loaded to the Status map.

What are your thoughts on the above?